File: frontend/src/components/benchmark/content/bellman_ford/implementations/bellman_ford.php
<?php

class BellmanFord {
    private array $edges = [];
    private array $vertices = [];

    public function addEdge(int $u, int $v, int $weight): void {
        $this->edges[] = ['u' => $u, 'v' => $v, 'weight' => $weight];
        $this->vertices[$u] = true;
        $this->vertices[$v] = true;
    }

    public function buildFromEdges(array $edgeList): void {
        foreach ($edgeList as $edge) {
            $this->addEdge($edge[0], $edge[1], $edge[2]);
        }
    }

    public function shortestPath(int $source): ?array {
        if (!isset($this->vertices[$source])) {
            return null;
        }

        $dist = [];
        foreach (array_keys($this->vertices) as $v) {
            $dist[$v] = INF;
        }
        $dist[$source] = 0;

        for ($i = 0; $i < count($this->vertices) - 1; $i++) {
            $updated = false;
            foreach ($this->edges as $edge) {
                if ($dist[$edge['u']] !== INF && $dist[$edge['u']] + $edge['weight'] < $dist[$edge['v']]) {
                    $dist[$edge['v']] = $dist[$edge['u']] + $edge['weight'];
                    $updated = true;
                }
            }
            if (!$updated) break;
        }

        foreach ($this->edges as $edge) {
            if ($dist[$edge['u']] !== INF && $dist[$edge['u']] + $edge['weight'] < $dist[$edge['v']]) {
                return null;
            }
        }

        return $dist;
    }

    public function hasNegativeCycle(): bool {
        $vertices = array_keys($this->vertices);
        return $this->shortestPath($vertices[0]) === null;
    }
}

?>

File: frontend/src/components/benchmark/content/bellman_ford/implementations/bellman_ford.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class BellmanFord
{
    private class Edge
    {
        public int U { get; set; }
        public int V { get; set; }
        public int Weight { get; set; }
    }

    private List<Edge> edges = new List<Edge>();
    private HashSet<int> vertices = new HashSet<int>();

    public void AddEdge(int u, int v, int weight)
    {
        edges.Add(new Edge { U = u, V = v, Weight = weight });
        vertices.Add(u);
        vertices.Add(v);
    }

    public void BuildFromEdges(List<(int, int, int)> edgeList)
    {
        foreach (var (u, v, w) in edgeList)
        {
            AddEdge(u, v, w);
        }
    }

    public Dictionary<int, int> ShortestPath(int source)
    {
        if (!vertices.Contains(source))
            return null;

        var dist = new Dictionary<int, int>();
        foreach (int v in vertices)
            dist[v] = int.MaxValue;
        dist[source] = 0;

        for (int i = 0; i < vertices.Count - 1; i++)
        {
            bool updated = false;
            foreach (var edge in edges)
            {
                if (dist[edge.U] != int.MaxValue && dist[edge.U] + edge.Weight < dist[edge.V])
                {
                    dist[edge.V] = dist[edge.U] + edge.Weight;
                    updated = true;
                }
            }
            if (!updated) break;
        }

        foreach (var edge in edges)
        {
            if (dist[edge.U] != int.MaxValue && dist[edge.U] + edge.Weight < dist[edge.V])
            {
                return null;
            }
        }

        return dist;
    }

    public bool HasNegativeCycle()
    {
        int start = vertices.First();
        return ShortestPath(start) == null;
    }
}